pkg/config/validation/validation.go

```diff
@@ -658,15 +667,82 @@ var ValidateEnvoyFilter = registerValidateFunc("ValidateEnvoyFilter",
 			} else {
 				// Run with strict validation, and emit warnings. This helps capture cases like unknown fields
 				// We do not want to reject in case the proto is valid but our libraries are outdated
-				if _, err := xds.BuildXDSObjectFromStruct(cp.ApplyTo, cp.Patch.Value, true); err != nil {
+				obj, err := xds.BuildXDSObjectFromStruct(cp.ApplyTo, cp.Patch.Value, true)
+				if err != nil {
 					errs = appendValidation(errs, WrapWarning(err))
 				}
+
+				// Append any deprecation notices
+				errs = appendValidation(errs, validateDeprecatedFilterTypes(obj))
 			}
 		}
 
 		return errs.Unwrap()
 	})
 
+func validateListenerMatchName(name string) error {
+	if newName, f := xds.ReverseDeprecatedFilterNames[name]; f {
+		return WrapWarning(fmt.Errorf("using deprecated filter name %q; use %q instead", name, newName))
+	}
+	return nil
+}
+
+func recurseDeprecatedTypes(message protoreflect.Message) ([]string, error) {
+	var topError error
+	var deprecatedTypes []string
+	if message == nil {
+		return nil, nil
+	}
+	message.Range(func(descriptor protoreflect.FieldDescriptor, value protoreflect.Value) bool {
+		m, isMessage := value.Interface().(protoreflect.Message)
+		if isMessage {
+			anyMessage, isAny := m.Interface().(*any.Any)
+			if isAny {
+				mt, err := protoregistry.GlobalTypes.FindMessageByURL(anyMessage.TypeUrl)
```

Comment:
Doesn't this depend on how many protos were loaded into istio? I think some istio filters, for example, may not have corresponding golang protos for C++ protos (because gogo)?

Reply:
We load every proto in go-control-plane (https://github.com/istio/istio/blob/19c8ea42031a5fb19c64654e6d8c9e468339460b/pkg/config/xds/filter_types.gen.go#L21). If they reference an unknown type, we would have gotten an error from a previous check and failed.

So I agree with the problem, but I think both the real EnvoyFilter logic would be broken in this scenario, and the existing validation logic already fails in this scenario.

Reply:
Since istio builds are likely to be customized based on how much of xDS they load (e.g. agent vs istiod), it might be better to pre-generate this list with make gen.